content/2020-05-22-book-review-writing-an-interpreter-in-go/index.md

```diff
@@ -0,0 +1,581 @@
+    - [top-down vs bottom-up vs recursive decent parsing](#top-down-vs-bottom-up-vs-recursive-decent-parsing)
+    - [binding vs identifier vs expression](#binding-vs-identifier-vs-expression)
+    - [AST with Nodes](#ast-with-nodes)
+    - [operator precedence](#operator-precedence)
+    - [prefix vs infix operators](#prefix-vs-infix-operators)
+    - [binary operators](#binary-operators)
+    - [Vaughan Pratt](#vaughan-pratt)
+    - [terminology](#terminology)
+    - [IntegerLiteral](#integerliteral)
+    - [prefixParseFns and infixParseFns](#prefixparsefns-and-infixparsefns)
+    - [parseBoolean](#parseboolean)
+    - [grouped expressions](#grouped-expressions)
+    - [block statements](#block-statements)
+    - [infix parse functions](#infix-parse-functions)
+    - [serialization](#serialization)
+  - [Chapter 3 - Evaluation](#chapter-3---evaluation)
+    - [interpreters vs compilers](#interpreters-vs-compilers)
+    - [tree-walking interpreters](#tree-walking-interpreters)
+    - [intermediate representation](#intermediate-representation)
+    - [just in time](#just-in-time)
+    - [host language](#host-language)
+    - [two instances of boolean](#two-instances-of-boolean)
+    - [unary operator expression](#unary-operator-expression)
+    - [eight infix operators](#eight-infix-operators)
+    - [comparing values directly](#comparing-values-directly)
+    - [return statements](#return-statements)
+    - [bindings](#bindings)
+    - [environment](#environment)
+    - [first-class citizens](#first-class-citizens)
+  - [Chapter 4 - Extending the Interpreter](#chapter-4---extending-the-interpreter)
+    - [built-in functions](#built-in-functions)
+    - [validate hash keys in evaluation stage](#validate-hash-keys-in-evaluation-stage)
+  - [Chapter 5 - Macro System](#chapter-5---macro-system)
+- [Overall Thoughts](#overall-thoughts)
+- [Recommend the Book?](#recommend-the-book)
+- [Purchase the Book](#purchase-the-book)
+
+## Notes
+
+This is a collection of my notes from each chapter in the book, along with my overall thoughts and recommendations.
+
+### Introduction
+
+In the beginning of the book, Thorsten explains why he wrote this book and who it's for. One thing that stuck out to me was this,
+
+> So I wrote this book, for you and me. This is the book I wish I had. This is a book for people who love to look under the hood. For people that love to learn by understanding how something really works.
+>
+> -Ball 7
+
+The part that resonated with me was the last sentence in hat quote -- "people that love to learn by understanding how something really works." That made me glad I picked this book up.
+
+> "Tree-walking interpreter"
+
+I highlighted this because I know I'll forgot what it means if I don't write it down. Summarized, it means an interpreter that parses the source code, builds an abstract syntax tree (AST) and then evaluates it. The "tree" comes from AST. The "walking" comes from walking down the tree.
```

Comment:
```suggestion
I highlighted this because I know I'll forget what it means if I don't write it down. Summarized, it means an interpreter that parses the source code, builds an abstract syntax tree (AST) and then evaluates it. The "tree" comes from AST. The "walking" comes from walking down the tree.
```